From 9a7e3bee486fd23ad32eecfb01fe8ed5fae41f96 Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Mon, 10 Jan 2022 08:05:54 +0000
Subject: [PATCH] xen/arm: mpam: Add MPAM Xen cmdline with related parsing for
 Dom0

As users may need to limit the cache/memory-bandwidth of Dom0, it
is required to provide a user interface to configure such information.

Considering this user interface should be applied to both ACPI and
device tree and since for ACPI, we could not update ACPI table to
embed Dom0 MPAM configuration. Therefore, a simple method to configure
the MPAM parameters, in order to limit the cache/mem-bandwidth for
Dom0 is using Xen cmdline.

This commit adds the MPAM Xen cmdline for Dom0 with related parsing
function and data structure. Currently, only add support for L3 cache.

Document the `dom0_mpam` cmdline should follow the format of:
```
dom0_mpam=l3:<L3_bitmap>
```

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-3762
---
 docs/misc/xen-command-line.pandoc | 11 +++++++++
 xen/arch/arm/include/asm/mpam.h   |  3 +++
 xen/arch/arm/mpam.c               | 37 +++++++++++++++++++++++++++++++
 xen/include/public/arch-arm.h     |  4 ++++
 4 files changed, 55 insertions(+)

diff --git a/docs/misc/xen-command-line.pandoc b/docs/misc/xen-command-line.pandoc
index 6b07d0f3a17f..5aeac07386b4 100644
--- a/docs/misc/xen-command-line.pandoc
+++ b/docs/misc/xen-command-line.pandoc
@@ -2864,3 +2864,14 @@ the hypervisor was compiled with `CONFIG_XSM` enabled.
 * `silo`: this will deny any unmediated communication channels between
   unprivileged VMs.  To choose this, the separated option in kconfig must also
   be enabled.
+
+### dom0_mpam (Arm64)
+> `= List of ( l3:<bitmap> )`
+
+Set the amount of MPAM resource limit for the initial domain (dom0). This is an
+optional command line option.
+
+* `l3:<bitmap>` specifies the limit of l3-cache. <bitmap> is a maximum 32bit
+bitmap. System l3 cache resource is divided into numbers of portions. Each
+portion corresponds to a bit in <bitmap> and the set bit represents the
+corresponding cache portion is available.
diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index d1ba5d7501fb..bf5f89953c88 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -9,6 +9,7 @@
 #include <xen/cpumask.h>
 #include <xen/spinlock.h>
 #include <asm/cpufeature.h>
+#include <public/arch-arm.h>
 
 /*
  * MPAM MSCs have the following register layout. See:
@@ -407,6 +408,8 @@ struct mpam_msc_ris
     struct mpam_component *comp;
 };
 
+extern struct xen_arch_domctl_mpam dom0_mpam_config;
+
 extern void mpam_postsmp_setup(void);
 extern void mpam_presmp_setup(void);
 extern int mpam_discovery_cpu_online(unsigned int cpu);
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 8c33dec479a4..2186e30b73e8 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -22,11 +22,14 @@
 #include <xen/acpi.h>
 #include <xen/device_tree.h>
 #include <xen/err.h>
+#include <xen/param.h>
+#include <xen/string.h>
 #include <xen/vmap.h>
 #include <asm/io.h>
 #include <asm/mpam.h>
 #include <asm/smp.h>
 
+struct xen_arch_domctl_mpam __initdata dom0_mpam_config;
 /*
  * mpam_list_lock protects the global lists when writing.
  * Once the MPAM is enabled these lists are read-only.
@@ -53,6 +56,40 @@ static uint8_t mpam_pmg_max;
 
 static bool __read_mostly mpam_enabled = false;
 
+static int __init parse_dom0_mpam_cmdline(const char *s)
+{
+    char *token, *cmdline = (char *)s;
+    const char *val;
+    uint64_t size;
+
+    if ( !cmdline )
+    {
+        printk(XENLOG_WARNING "MPAM: No Dom0 MPAM configuration found\n");
+        return -ENOENT;
+    }
+
+    token = strsep(&cmdline, ",");
+    while ( token )
+    {
+        val = strchr(token, ':') + 1;
+        size = parse_size_and_unit(val, &val);
+
+        if ( strncmp(token, "l3", strlen("l3")) == 0 )
+            /* parse_size_and_unit() default returns KB if no unit specified */
+            dom0_mpam_config.l3_cpbm = size >>= 10;
+        else
+        {
+            printk(XENLOG_ERR "MPAM: Invalid Dom0 MPAM cmdline\n");
+            return -EINVAL;
+        }
+
+        token = strsep(&cmdline, ",");
+    }
+
+    return 0;
+}
+custom_param("dom0_mpam", parse_dom0_mpam_cmdline);
+
 static struct dt_device_node *
 __init dt_find_next_cache_node(const struct dt_device_node *np)
 {
diff --git a/xen/include/public/arch-arm.h b/xen/include/public/arch-arm.h
index a25e87dbda3a..598107e2f756 100644
--- a/xen/include/public/arch-arm.h
+++ b/xen/include/public/arch-arm.h
@@ -327,6 +327,10 @@ DEFINE_XEN_GUEST_HANDLE(vcpu_guest_context_t);
 #define XEN_DOMCTL_CONFIG_TEE_OPTEE     1
 #define XEN_DOMCTL_CONFIG_TEE_FFA       2
 
+struct xen_arch_domctl_mpam {
+    uint32_t l3_cpbm;
+};
+
 struct xen_arch_domainconfig {
     /* IN/OUT */
     uint8_t gic_version;
-- 
2.34.1

